modin/engines/base/frame/data.py

```diff
@@ -1746,26 +1733,38 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
 
         index_other_obj = [o.axes[axis] for o in other]
         joined_index = self._join_index_objects(axis, index_other_obj, how, sort)
+
         # We have to set these because otherwise when we perform the functions it may
         # end up serializing this entire object.
         left_old_idx = self.axes[axis]
         right_old_idxes = index_other_obj
 
-        def make_map_func():
-            if not joined_index.is_unique and axis == 0:
+        def make_map_func(index, left=True):
+            # left - specific argument for case of binary operation;
+            # it choose indexer for left or right index
+            if index.equals(joined_index):
                 return lambda df: df
+            if reindexer == "binary":
+                # case for binary operation with duplicate values; way from pandas
+                _join_index, ilidx, iridx = self.axes[axis].join(
```

Comment:
in that case, we're doing double joining here, once for `joined_index` and the second here, can we avoid first joining if we don't need it?